src/shared/tests.c

```
@@ -42,5 +42,9 @@ const char* get_exe_relative_testdata_dir(void) {
 
         assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
         assert_se(snprintf(testdir, sizeof(testdir), "%s/testdata", dirname(exedir)) > 0);
+        if (access(testdir, F_OK) < 0) {
+                fprintf(stderr,  "Test data directory '%s' does not exist, set $TEST_DIR\n", testdir);
+                exit(1);
```

`exit(1)` prevents all cleanup actions.

For example, `test_engine` creates the directory:
```c

        assert_se(runtime_dir = setup_fake_runtime_dir());

        /* prepare the test */
        assert_se(set_unit_path(TEST_DATA_DIR("")) >= 0);
```
I got many `/tmp/fake-xdg-runtime-*`-directories :-)

Maybe, we should call `TEST_DATA_DIR` before `setup_fake_runtime_dir`. As another option we can propagate error from `get_exe_relative_testdata_dir`

---

Or just write it to stderr to get it as first error message, and let the test fail by itself later?

---

Yeah, doing `set_unit_path` before any "real" work seems best.

---

> and let the test fail by itself later?

our tests fail via `ABRT`. this prevents all cleanup operations too.

> Or just write it to stderr to get it as first error message

I think
```
$ ./libtool --mode=execute ./test-engine
Test data directory '/home/vagrant/systemd/.libs/testdata' does not exist, set $TEST_DIR
```
is better then
```
some stdout
...
Test data directory '/home/vagrant/systemd/.libs/testdata' does not exist, set $TEST_DIR
many lines
aborted
```

---

OK, agreed. I swapped around the setup of test dir and fake runtime dir in the four affected tests. Will push in a minute.